include/storage/io.hpp

```diff
@@ -20,11 +24,134 @@ namespace storage
+        {
+            throw util::exception("Fingerprint mismatch in " + filename);
+        }
+    }
+
+    /* Read count objects of type T into pointer dest */
+    template <typename T> void readInto(T *dest, const std::size_t count)
+    {
+        static_assert(std::is_trivially_copyable<T>::value,
+                      "bytewise reading requires trivially copyable type");
+        if (count == 0)
+            return;
+        input_stream.read(reinterpret_cast<char *>(dest), count * sizeof(T));
+
+        // safe to cast here, according to CPP docs, negative values for gcount
+        // are never used.
+        const unsigned long bytes_read = static_cast<unsigned long>(input_stream.gcount());
+        const auto expected_bytes = count * sizeof(T);
+
+        if (bytes_read == 0)
+        {
+            throw util::exception("Error reading from " + filename + ": " + std::strerror(errno));
+        }
+        else if (bytes_read < expected_bytes)
+        {
+            throw util::exception("Error reading from " + filename + ": Unexpected end of file");
+        }
+    }
+
+    template <typename T> void readInto(T &target) { readInto(&target, 1); }
+
+    template <typename T> T readOne()
+    {
+        T tmp;
+        readInto(tmp);
+        return tmp;
+    }
+
+    template <typename T> void skip(const std::size_t element_count)
+    {
+        boost::iostreams::seek(input_stream, element_count * sizeof(T), BOOST_IOS::cur);
+    }
+
+    /*******************************************/
+
+    std::uint32_t readElementCount32() { return readOne<std::uint32_t>(); }
+    std::uint64_t readElementCount64() { return readOne<std::uint64_t>(); }
+
+    template <typename T> void deserializeVector(std::vector<T> &data)
+    {
+        const auto count = readElementCount64();
+        data.resize(count);
+        readInto(data.data(), count);
+    }
+
+    bool readAndCheckFingerprint()
+    {
+        auto fingerprint = readOne<util::FingerPrint>();
+        const auto valid = util::FingerPrint::GetValid();
+        // compare the compilation state stored in the fingerprint
+        return valid.IsMagicNumberOK(fingerprint) && valid.TestContractor(fingerprint) &&
+               valid.TestGraphUtil(fingerprint) && valid.TestRTree(fingerprint) &&
+               valid.TestQueryObjects(fingerprint);
+    }
+
+    std::size_t size()
+    {
+        auto current_pos = input_stream.tellg();
+        input_stream.seekg(0, input_stream.end);
+        auto length = input_stream.tellg();
+        input_stream.seekg(current_pos, input_stream.beg);
+        return length;
+    }
+
+    std::vector<std::string> readLines()
+    {
+        std::vector<std::string> result;
+        std::string thisline;
+        try
+        {
+            while (std::getline(input_stream, thisline))
+            {
+                std::clog << "Read " << thisline << std::endl;
+                result.push_back(thisline);
```

Comment:
Can't move here - we're re-using the same string object on each iteration of the loop.